### PR TITLE
Detached QPs and hidden replies

### PR DESCRIPTION
**Pull request by estrattonbailey**
Adds quote gates, detached quotes, and hidden replies.

### CODE REVIEW

File: src/components/icons/X.tsx
Comment (estrattonbailey) [inline src/components/icons/X.tsx:1]: Silly, will remove

File: src/state/queries/postgate/index.ts [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+          repo: urip.host,
+          collection: POSTGATE_COLLECTION,
+          rkey: urip.rkey,
+        }),
+    )
+
+    if (data.value && AppBskyFeedPostgate.isRecord(data.value)) {
+      return data.value
+    } else {
+      return undefined
```
Comment (estrattonbailey) [inline src/state/queries/postgate/index.ts:66]: Could return null here instead (and at threadgate) and simplify some type checks

**Comment by haileyok** [comment]
So far things are looking good:

- Had to push a commit to fix toast import, just be sure to pull next time you check out
- Wonder if we want to put your own reply behind `Show hidden replies` whenever the OP creator hides your post. I don't necessarily think that's a problem - in fact maybe that's preferred - but just something to think about
- The `Hide post for me` button might better if we say `Hide reply for me`, so it lines up with `Hide reply for everyone`
- Should check how difficult it is to either:
  - Have the respost/quote dialog use the new ALF system (swear we did this)
  - Make the disabled button grayed out

**Comment by haileyok** [comment]
> @estrattonbailey
[Use a set for faster lookup](https://github.com/bluesky-social/social-app/pull/4878/commits/dc11707e54615262165b1c6f2e3b5f1e7d80bb89)

great minds think alike 🍻

File: src/view/com/composer/threadgate/ThreadgateBtn.tsx [hunk 2]
```diff
@@ -59,16 +67,19 @@ export function ThreadgateBtn({
...
         </Button>
       </Animated.View>
-      <ThreadgateEditorDialog
+      <PostInteractionSettingsControlledDialog
         control={control}
-        threadgate={threadgate}
-        onChange={onChange}
+        onSave={() => {
+          control.close()
+        }}
```
Comment (haileyok) [inline src/view/com/composer/threadgate/ThreadgateBtn.tsx:78]: these always suss me out. i think its fine but lets make sure it doesnt break android back button.

File: src/view/com/post-thread/PostThreadItem.tsx [hunk 3]
```diff
@@ -77,7 +80,9 @@ export function PostThreadItem({
   overrideBlur: boolean
   onPostReply: (postUri: string | undefined) => void
   hideTopBorder?: boolean
+  threadgateRecord?: AppBskyFeedThreadgate.Record
 }) {
+  // TODO should load this for every post, ugh
```
Comment (haileyok) [inline src/view/com/post-thread/PostThreadItem.tsx:85]: just making so we dont forget

File: src/view/com/posts/FeedItem.tsx [hunk 4]
```diff
@@ -381,23 +400,63 @@ let FeedItemInner = ({
...
   const pal = usePalette('default')
   const {_} = useLingui()
+  const {currentAccount} = useSession()
   const [limitLines, setLimitLines] = useState(
     () => countLines(richText.text) >= MAX_POST_LINES,
   )
+  const {uris: hiddenReplyUris, recentlyUnhiddenUris} =
+    useThreadgateHiddenReplyUris()
+  const additionalPostAlerts: AppModerationCause[] = React.useMemo(() => {
+    const isPostHiddenByHiddenReplyCache = hiddenReplyUris.includes(post.uri)
```
Comment (haileyok) [inline src/view/com/posts/FeedItem.tsx:428]: didn't see where this comes from but can this be a set too?

File: src/view/com/util/forms/PostDropdownBtn.tsx [hunk 5]
```diff
@@ -104,17 +119,46 @@ let PostDropdownBtn = ({
...
+  const {uris: hiddenReplies, recentlyUnhiddenUris} =
+    useThreadgateHiddenReplyUris()
+
   const postUri = post.uri
   const postCid = post.cid
   const postAuthor = post.author
+  const quoteEmbed = React.useMemo(() => {
+    if (!currentAccount || !post.embed) return
+    return getMaybeDetachedQuoteEmbed({
+      viewerDid: currentAccount?.did,
```
Comment (haileyok) [inline src/view/com/util/forms/PostDropdownBtn.tsx:136]: no need for `?` here, or we can remove `!currentAccount` above

File: src/view/com/posts/FeedItem.tsx (hunk 4 above)
Comment (estrattonbailey) [reply src/view/com/posts/FeedItem.tsx:428]: Oh yeah good call, might as well. It's an account-scoped context value, for optimistic state.

File: src/lib/api/index.ts [hunk 6]
```diff
@@ -244,54 +253,43 @@ export async function post(agent: BskyAgent, opts: PostOpts) {
 
   try {
     // TODO: this needs to be batch-created with the post!
```
[the remark this answers is not in the document]
Comment (estrattonbailey) [reply src/lib/api/index.ts:255]: I think we'd be looking at doing an `applyWrites` call, like in the thread composer PR. Would be a good win, but we don't need to address it right now. The user is notified if these writes fail, and they can edit the gating immediately after posting.

File: src/lib/moderation/useModerationCauseDescription.ts [hunk 7]
```diff
@@ -111,6 +114,18 @@ export function useModerationCauseDescription(
         description: _(msg`You have hidden this post`),
       }
     }
+    if (cause.type === 'reply-hidden') {
+      const isYou = currentAccount?.did === cause.source.did
```
[the remark this answers is not in the document]
Comment (estrattonbailey) [reply src/lib/moderation/useModerationCauseDescription.ts:118]: heh yeah totally

File: src/lib/api/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (haileyok) [reply src/lib/api/index.ts:255]: ah nice, makes sense.

File: src/view/com/post-thread/PostThreadItem.tsx (hunk 3 above)
Comment (estrattonbailey) [reply src/view/com/post-thread/PostThreadItem.tsx:85]: oh lawl this is old, long done

File: src/view/com/composer/threadgate/ThreadgateBtn.tsx (hunk 2 above)
Comment (estrattonbailey) [reply src/view/com/composer/threadgate/ThreadgateBtn.tsx:78]: Yep cool, let's leave this one unresolved for now

File: src/components/dialogs/PostInteractionSettingsDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+  style?: StyleProp<ViewStyle>
+  disabled?: boolean
+}) {
+  const t = useTheme()
+  return (
+    <Button
+      disabled={disabled}
+      onPress={onPress}
+      label={label}
+      accessibilityHint="Select this option"
```
Comment (surfdude29) [inline src/components/dialogs/PostInteractionSettingsDialog.tsx:471]: Would it be helpful to have this `accessibilityHint` marked for localization? (Tbh I'm unclear on what the role of `accessibilityHint`s is generally which is why I ask :)

File: src/lib/api/index.ts [hunk 9]
```diff
@@ -244,54 +253,43 @@ export async function post(agent: BskyAgent, opts: PostOpts) {
...
+    })
   } catch (e: any) {
-    console.error(`Failed to create threadgate: ${e.toString()}`)
+    logger.error(`Failed to create threadgate`, {
+      context: 'composer',
+      safeMessage: e.message,
+    })
     throw new Error(
-      'Post reply-controls failed to be set. Your post was created but anyone can reply to it.',
+      'Failed to save post interaction settings. Your post was created but users may be able to interact with it.',
```
Comment (surfdude29) [inline src/lib/api/index.ts:270]: Would it be feasible to mark this error string for localization?

File: src/lib/api/index.ts [hunk 10]
```diff
@@ -244,54 +253,43 @@ export async function post(agent: BskyAgent, opts: PostOpts) {
...
+        post: res.uri,
+      },
+    })
+  } catch (e: any) {
+    logger.error(`Failed to create postgate`, {
+      context: 'composer',
+      safeMessage: e.message,
+    })
+    throw new Error(
+      'Failed to save post interaction settings. Your post was created but users may be able to interact with it.',
```
Comment (surfdude29) [inline src/lib/api/index.ts:290]: Same question here

File: src/components/WhoCanReply.tsx [hunk 11]
```diff
@@ -186,79 +160,84 @@ function WhoCanReplyDialog({
...
+  settings: ThreadgateAllowUISetting[]
+  embeddingDisabled: boolean
 }) {
+  const {_} = useLingui()
   return (
     <Dialog.Outer control={control}>
       <Dialog.Handle />
-      <WhoCanReplyDialogInner post={post} settings={settings} />
+      <Dialog.ScrollableInner
+        label={_(msg`Who can reply dialog`)}
```
[the remark this answers is not in the document]
Comment (estrattonbailey) [reply src/components/WhoCanReply.tsx:175]: Ah yeah we can do better here, good call

File: src/lib/api/index.ts (hunk 9 above)
Comment (estrattonbailey) [reply src/lib/api/index.ts:270]: Love that suggestion. I'll make a separate PR to clean some of this up and include these once this PR is in 👍

**Comment by haileyok** [comment]
Lets

Go

🥳

**Review by haileyok — APPROVED** [review]

File: src/lib/api/index.ts (hunk 9 above)
Comment (surfdude29) [reply src/lib/api/index.ts]: @estrattonbailey Afaik the user-facing strings in `src/lib/api/index.ts` are some of the only ones in the app that can't yet be localized, so when you've got a moment sometime that PR would be appreciated :)

**Pull request merged**